### ISSUE TITLE
Docker registry does not work behind TLS-only proxy (stunnel)

### ISSUE THREAD

**Issue by lwcolton** [docker/distribution#996]
I have docker registry set up behind stunnel.  I could use curl to GET /v2/ just fine, but for some reason image pushes would fail.   I discovered if the registry config does not have any tls settings under the http section, the registry assumes the client is communicating via http and responds to POST /v2/<name>/blobs/uploads with a URL prefix of http://   This causes the client to then attempt to push the image to the http:// url, and the world breaks.

Example:

curl -D - -X POST https://our.docker.server:8443/v2/gunicorn-base/blobs/uploads/
HTTP/1.1 202 Accepted
Content-Length: 0
Docker-Distribution-Api-Version: registry/2.0
Docker-Upload-Uuid: 4299c757-b2d3-40d2-bc9b-7cac8083c2ae
Location: http://our.docker.server/v2/gunicorn-base/blobs/uploads/<redacted>

Notice how the client originally requested a URL beginning with https:// and the server responded with a location starting with http://

My recommendation is to add a setting under the http or tls section that will tell the registry to use an https:// url prefix even though it does not have tls configured.
```
Infos:
# docker version
Client:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   d2e7ba4
 Built:        Thu Sep 17 23:46:39 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-dev
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   d2e7ba4
 Built:        Thu Sep 17 23:46:39 UTC 2015
 OS/Arch:      linux/amd64

# docker info
Containers: 9
Images: 637
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: xfs
 Dirs: 658
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-61-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 2
Total Memory: 3.908 GiB
Name: raptorcage
ID: R7QM:23DE:XOID:O2UY:JZCI:5BT3:A2WX:7C2F:M6DF:B4E3:XAPF:AZIJ
WARNING: No swap limit support

# docker exec registry_main_1 registry --version
registry github.com/docker/distribution v2.1.1

Registry Config:

version: 0.1
log:
  fields:
    service: registry
storage:
    cache:
        blobdescriptor: inmemory
    s3:
        accesskey: <redacted>
        secretkey: <redacted>
        region: us-east-1
        bucket: some-bucket
        rootdirectory: /v2-registry
        encrypt: true
        secure: true
        v4auth: true
http:
    addr: :5000
    headers:
        X-Content-Type-Options: [nosniff]
health:
  storagedriver:
    enabled: true

What happens when I push:
08eefe93c719: Pushing 1.024 kB
read tcp <redacted>:8443: connection reset by peer

What I see in the  syslog for stunnel (clearly indicating the docker client is trying to use http after seeing the bad referred send by the registry):

Sep 18 04:04:51 <redacted> stunnel: LOG5[10682:140222830589696]: Service [registry] accepted connection from <redacted>
Sep 18 04:04:51 <redacted> stunnel: LOG3[10682:140222830589696]: SSL_accept: 140760FC: error:140760FC:SSL routines:SSL23_GET_CLIENT_HELLO:unknown protocol
Sep 18 04:04:51 <redacted> stunnel: LOG5[10682:140222830589696]: Connection reset: 0 byte(s) sent to SSL, 0 byte(s) sent to socket
```

**Comment by stevvooe** [comment]
The issue is that the registry is not properly detecting the protocol since stunnel, operating at the tcp stream level, does not set the `X-Forwarded-Proto` or `X-Forwarded-Host` header. In addition, the registry cannot detect the correct url.

We will need a way to set the host for registry redirect responses.

@aaronlehmann @RichardScothern I propose we add a `host` parameter to be used for redirect responses when the URL cannot be correctly detected:

```yaml
http:
  addr: :5000
  host: https://our.docker.server:8443
```

The other option is to ensure the client correctly handles relative URLs from the `Location` field. My guess is that it already does and we can start issuing relative urls for these redirects to avoid these issues.

**Comment by lwcolton** [comment]
+1 for resolution via config setting for the Location field

On Fri, Sep 18, 2015 at 5:01 AM, Stephen Day <notifications@github.com>
wrote:

> The issue is that the registry is not properly detecting the protocol
> since stunnel, operating at the tcp stream level, does not set the
> X-Forwarded-Proto or X-Forwarded-Host header. In addition, the registry
> cannot detect the correct url.
>
> We will need a way to set the host for registry redirect responses.
>
> @aaronlehmann <https://github.com/aaronlehmann> @RichardScothern
> <https://github.com/RichardScothern> I propose we add a host parameter to
> be used for redirect responses when the URL cannot be correctly detected:
>
> http:
>   addr: :5000
>   host: https://our.docker.server:8443
>
> The other option is to ensure the client correctly handles relative URLs
> from the Location field. My guess is that it already does and we can
> start issuing relative urls for these redirects to avoid these issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/distribution/issues/996#issuecomment-141408919>
> .
>



-- 
Sincerely,

Colton Leekley-Winslow
651-242-9559
lwcolton@gmail.com

**Comment by aaronlehmann** [comment]
I like the idea of having a `host` parameter. I think this is pretty standard practice for HTTP servers (i.e. `ServerName` in Apache). It will give us more flexibility in case we find other situations where a proxy interferes with URL generation.

**Comment by djenriquez** [comment]
I believe I'm having the same issue with #1043, however I'm not really sure how to test this. I am able to pull images and push  UNCHANGED images, but cannot push new images or images with changes.

My private registry is behind an ELB, which handles the TLS termination then routed behind an NGINX reverse-proxy to the container. I also have an authentication server using [cesanta/docker_auth](https://github.com/cesanta/docker_auth) where traffic is TCP streamed from an ELB into the container, which then handles the TLS termination. Cesanta/docker_auth requires the TLS termination to happen in the container, which is why I'm not being consistent with TLS termination.

From what @lwcolton stated, since my registry config does not have TLS defined, it would push to an http:// location and would break.

Am I having the same issue? If so, why does it only affect pushes? Is there an ETA for a fix?

**Comment by stevvooe** [comment]
This should be addressed by #1035. It is only affecting pushes because redirects are used to assign upload locations.

**Comment by dmp42** [comment]
Closing since #1035 was merged.
